{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "This IPython notebook will walk through the steps of characterizing iModulons through the semi-automated tools in PyModulon. You will need:\n",
    "\n",
    "* M and A matrices\n",
    "* Expression data (e.g. `log_tpm_norm.csv`)\n",
    "* Gene table and KEGG/GO annotations (Generated in `gene_annotation.ipynb`)\n",
    "* Sample table, with a column for `project` and `condition`\n",
    "* TRN file\n",
    "\n",
    "Optional:\n",
    "* iModulon table (if you already have some characterized iModulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.core import IcaData\n",
    "from pymodulon.plotting import *\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import re\n",
    "from Bio.KEGG import REST\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = path.join('..','Species','Synechocystis_sp_PCC_6803')\n",
    "\n",
    "# Enter the location of your data here\n",
    "data_dir = path.join(project_dir,'computed_data')\n",
    "\n",
    "# GO and KEGG annotations are in the 'external' folder\n",
    "external_data = path.join(project_dir,'external')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your sample table (i.e. metadata file)\n",
    "Your metadata file will probably have a lot of columns, most of which you may not care about. Feel free to save a secondary copy of your metadata file with only columns that seem relevant to you. The two most important columns are:\n",
    "1. `project`\n",
    "2. `condition`\n",
    "\n",
    "Make sure that these columns exist in your metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRX12668170</th>\n",
       "      <td>thermal</td>\n",
       "      <td>Control_30_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX12668171</th>\n",
       "      <td>thermal</td>\n",
       "      <td>Control_30_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX12668172</th>\n",
       "      <td>thermal</td>\n",
       "      <td>Control_30_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX12668179</th>\n",
       "      <td>thermal</td>\n",
       "      <td>Control_30_9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX12668183</th>\n",
       "      <td>thermal</td>\n",
       "      <td>Control_30_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             project     condition\n",
       "SRX12668170  thermal  Control_30_0\n",
       "SRX12668171  thermal  Control_30_1\n",
       "SRX12668172  thermal  Control_30_2\n",
       "SRX12668179  thermal  Control_30_9\n",
       "SRX12668183  thermal  Control_30_1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata = pd.read_csv(path.join(data_dir,'final_metadata.tsv'),index_col=0,sep='\\t')\n",
    "df_metadata[['project','condition']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_metadata.project.notnull().all())\n",
    "print(df_metadata.condition.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your TRN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row of the TRN file represents a regulatory interaction.  \n",
    "**Your TRN file must have the following columns:**\n",
    "1. `regulator` - Name of regulator (`/` or `+` characters will be converted to `;`)\n",
    "1. `gene_id` - Locus tag of gene being regulated\n",
    "\n",
    "The following columns are optional, but are helpful to have:\n",
    "1. `regulator_id` - Locus tag of regulator\n",
    "1. `gene_name` - Name of gene (can automatically update this using `name2num`)\n",
    "1. `direction` - Direction of regulation ('+' for activation, '-' for repression, '?' or NaN for unknown)\n",
    "1. `evidence` - Evidence of regulation (e.g. ChIP-exo, qRT-PCR, SELEX, Motif search)\n",
    "1. `PMID` - Reference for regulation\n",
    "\n",
    "You may add any other columns that could help you. TRNs may be saved as either CSV or TSV files. See below for an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../Species/Synechocystis_sp_PCC_6803/external/TRN.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_19186/1170489286.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_trn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexternal_data\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'TRN.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mdf_trn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    309\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m                 )\n\u001b[0;32m--> 311\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    313\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    584\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    585\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 586\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    587\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    588\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    480\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    481\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 482\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    483\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    484\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    809\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    810\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 811\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    812\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    813\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1038\u001b[0m             )\n\u001b[1;32m   1039\u001b[0m         \u001b[0;31m# error: Too many arguments for \"ParserBase\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1040\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mmapping\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[call-arg]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1041\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1042\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_failover_to_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/io/parsers/c_parser_wrapper.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m     49\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     50\u001b[0m         \u001b[0;31m# open handles\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 51\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_open_handles\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     52\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/io/parsers/base_parser.py\u001b[0m in \u001b[0;36m_open_handles\u001b[0;34m(self, src, kwds)\u001b[0m\n\u001b[1;32m    227\u001b[0m             \u001b[0mmemory_map\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"memory_map\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    228\u001b[0m             \u001b[0mstorage_options\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"storage_options\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 229\u001b[0;31m             \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"encoding_errors\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"strict\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    230\u001b[0m         )\n\u001b[1;32m    231\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/io/common.py\u001b[0m in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    705\u001b[0m                 \u001b[0mencoding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    706\u001b[0m                 \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 707\u001b[0;31m                 \u001b[0mnewline\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    708\u001b[0m             )\n\u001b[1;32m    709\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../Species/Synechocystis_sp_PCC_6803/external/TRN.csv'"
     ]
    }
   ],
   "source": [
    "df_trn = pd.read_csv(path.join(external_data,'TRN.csv'))\n",
    "df_trn.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `regulator` and `gene_id` must be filled in for each row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_trn' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_77172/242824108.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_trn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregulator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnotnull\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_trn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgene_id\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnotnull\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df_trn' is not defined"
     ]
    }
   ],
   "source": [
    "print(df_trn.regulator.notnull().all())\n",
    "print(df_trn.gene_id.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the data\n",
    "You're now ready to load your IcaData object!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>operon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SGL_RS01370</td>\n",
       "      <td>SGL_RS01370</td>\n",
       "      <td>NC_005232.1</td>\n",
       "      <td>slr6001</td>\n",
       "      <td>243</td>\n",
       "      <td>3053</td>\n",
       "      <td>+</td>\n",
       "      <td>hybrid sensor histidine kinase/response regulator</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SGL_RS00010</td>\n",
       "      <td>SGL_RS00010</td>\n",
       "      <td>NC_005229.1</td>\n",
       "      <td>sll5002</td>\n",
       "      <td>574</td>\n",
       "      <td>1017</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SGL_RS00635</td>\n",
       "      <td>SGL_RS00635</td>\n",
       "      <td>NC_005230.1</td>\n",
       "      <td>sll7002</td>\n",
       "      <td>706</td>\n",
       "      <td>1287</td>\n",
       "      <td>-</td>\n",
       "      <td>IS701 family transposase</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op2588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SGL_RS01880</td>\n",
       "      <td>SGL_RS01880</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0612</td>\n",
       "      <td>811</td>\n",
       "      <td>1494</td>\n",
       "      <td>+</td>\n",
       "      <td>pseudouridine synthase</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SGL_RS00015</td>\n",
       "      <td>SGL_RS00015</td>\n",
       "      <td>NC_005229.1</td>\n",
       "      <td>sll5003</td>\n",
       "      <td>1136</td>\n",
       "      <td>1540</td>\n",
       "      <td>-</td>\n",
       "      <td>DUF5615 family PIN-like protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3737</th>\n",
       "      <td>SGL_RS18370</td>\n",
       "      <td>SGL_RS18370</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0607</td>\n",
       "      <td>3569134</td>\n",
       "      <td>3569523</td>\n",
       "      <td>+</td>\n",
       "      <td>cyclic nucleotide-binding domain-containing pr...</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3738</th>\n",
       "      <td>SGL_RS18375</td>\n",
       "      <td>hisIE</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0608</td>\n",
       "      <td>3569672</td>\n",
       "      <td>3570319</td>\n",
       "      <td>+</td>\n",
       "      <td>bifunctional phosphoribosyl-AMP cyclohydrolase...</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>A0A6P1VM37_9SYNC</td>\n",
       "      <td>Op2513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3739</th>\n",
       "      <td>SGL_RS18380</td>\n",
       "      <td>SGL_RS18380</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0609</td>\n",
       "      <td>3570424</td>\n",
       "      <td>3571575</td>\n",
       "      <td>+</td>\n",
       "      <td>GTP-binding protein</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3740</th>\n",
       "      <td>SGL_RS18385</td>\n",
       "      <td>SGL_RS18385</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0610</td>\n",
       "      <td>3571711</td>\n",
       "      <td>3572403</td>\n",
       "      <td>+</td>\n",
       "      <td>ABC transporter permease</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3741</th>\n",
       "      <td>SGL_RS01875</td>\n",
       "      <td>sds</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0611</td>\n",
       "      <td>3573271</td>\n",
       "      <td>3574242</td>\n",
       "      <td>+</td>\n",
       "      <td>solanesyl diphosphate synthase</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>A0A6P1VKP0_9SYNC</td>\n",
       "      <td>Op1759</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3742 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        locus_tag    gene_name    accession old_locus_tag    start      end  \\\n",
       "0     SGL_RS01370  SGL_RS01370  NC_005232.1       slr6001      243     3053   \n",
       "1     SGL_RS00010  SGL_RS00010  NC_005229.1       sll5002      574     1017   \n",
       "2     SGL_RS00635  SGL_RS00635  NC_005230.1       sll7002      706     1287   \n",
       "3     SGL_RS01880  SGL_RS01880  NC_000911.1       slr0612      811     1494   \n",
       "4     SGL_RS00015  SGL_RS00015  NC_005229.1       sll5003     1136     1540   \n",
       "...           ...          ...          ...           ...      ...      ...   \n",
       "3737  SGL_RS18370  SGL_RS18370  NC_000911.1       slr0607  3569134  3569523   \n",
       "3738  SGL_RS18375        hisIE  NC_000911.1       slr0608  3569672  3570319   \n",
       "3739  SGL_RS18380  SGL_RS18380  NC_000911.1       slr0609  3570424  3571575   \n",
       "3740  SGL_RS18385  SGL_RS18385  NC_000911.1       slr0610  3571711  3572403   \n",
       "3741  SGL_RS01875          sds  NC_000911.1       slr0611  3573271  3574242   \n",
       "\n",
       "     strand                                       gene_product  \\\n",
       "0         +  hybrid sensor histidine kinase/response regulator   \n",
       "1         -                               hypothetical protein   \n",
       "2         -                           IS701 family transposase   \n",
       "3         +                             pseudouridine synthase   \n",
       "4         -                    DUF5615 family PIN-like protein   \n",
       "...     ...                                                ...   \n",
       "3737      +  cyclic nucleotide-binding domain-containing pr...   \n",
       "3738      +  bifunctional phosphoribosyl-AMP cyclohydrolase...   \n",
       "3739      +                                GTP-binding protein   \n",
       "3740      +                           ABC transporter permease   \n",
       "3741      +                     solanesyl diphosphate synthase   \n",
       "\n",
       "                                                  COG           uniprot  \\\n",
       "0                      Signal transduction mechanisms               NaN   \n",
       "1                                   No COG annotation               NaN   \n",
       "2               Replication, recombination and repair               NaN   \n",
       "3     Translation, ribosomal structure and biogenesis               NaN   \n",
       "4                                   No COG annotation               NaN   \n",
       "...                                               ...               ...   \n",
       "3737                   Signal transduction mechanisms               NaN   \n",
       "3738              Nucleotide transport and metabolism  A0A6P1VM37_9SYNC   \n",
       "3739                                 Function unknown               NaN   \n",
       "3740                                 Function unknown               NaN   \n",
       "3741                Coenzyme transport and metabolism  A0A6P1VKP0_9SYNC   \n",
       "\n",
       "      operon  \n",
       "0       Op97  \n",
       "1      Op712  \n",
       "2     Op2588  \n",
       "3      Op465  \n",
       "4      Op686  \n",
       "...      ...  \n",
       "3737   Op766  \n",
       "3738  Op2513  \n",
       "3739   Op489  \n",
       "3740   Op108  \n",
       "3741  Op1759  \n",
       "\n",
       "[3742 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table = pd.read_csv(path.join(data_dir,'gene_info.csv'))\n",
    "gene_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>operon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SGL_RS01370</th>\n",
       "      <td>SGL_RS01370</td>\n",
       "      <td>NC_005232.1</td>\n",
       "      <td>slr6001</td>\n",
       "      <td>243</td>\n",
       "      <td>3053</td>\n",
       "      <td>+</td>\n",
       "      <td>hybrid sensor histidine kinase/response regulator</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SGL_RS00010</th>\n",
       "      <td>SGL_RS00010</td>\n",
       "      <td>NC_005229.1</td>\n",
       "      <td>sll5002</td>\n",
       "      <td>574</td>\n",
       "      <td>1017</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SGL_RS00635</th>\n",
       "      <td>SGL_RS00635</td>\n",
       "      <td>NC_005230.1</td>\n",
       "      <td>sll7002</td>\n",
       "      <td>706</td>\n",
       "      <td>1287</td>\n",
       "      <td>-</td>\n",
       "      <td>IS701 family transposase</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op2588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SGL_RS01880</th>\n",
       "      <td>SGL_RS01880</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0612</td>\n",
       "      <td>811</td>\n",
       "      <td>1494</td>\n",
       "      <td>+</td>\n",
       "      <td>pseudouridine synthase</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SGL_RS00015</th>\n",
       "      <td>SGL_RS00015</td>\n",
       "      <td>NC_005229.1</td>\n",
       "      <td>sll5003</td>\n",
       "      <td>1136</td>\n",
       "      <td>1540</td>\n",
       "      <td>-</td>\n",
       "      <td>DUF5615 family PIN-like protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op686</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_name    accession old_locus_tag  start   end strand  \\\n",
       "locus_tag                                                                 \n",
       "SGL_RS01370  SGL_RS01370  NC_005232.1       slr6001    243  3053      +   \n",
       "SGL_RS00010  SGL_RS00010  NC_005229.1       sll5002    574  1017      -   \n",
       "SGL_RS00635  SGL_RS00635  NC_005230.1       sll7002    706  1287      -   \n",
       "SGL_RS01880  SGL_RS01880  NC_000911.1       slr0612    811  1494      +   \n",
       "SGL_RS00015  SGL_RS00015  NC_005229.1       sll5003   1136  1540      -   \n",
       "\n",
       "                                                  gene_product  \\\n",
       "locus_tag                                                        \n",
       "SGL_RS01370  hybrid sensor histidine kinase/response regulator   \n",
       "SGL_RS00010                               hypothetical protein   \n",
       "SGL_RS00635                           IS701 family transposase   \n",
       "SGL_RS01880                             pseudouridine synthase   \n",
       "SGL_RS00015                    DUF5615 family PIN-like protein   \n",
       "\n",
       "                                                         COG uniprot  operon  \n",
       "locus_tag                                                                     \n",
       "SGL_RS01370                   Signal transduction mechanisms     NaN    Op97  \n",
       "SGL_RS00010                                No COG annotation     NaN   Op712  \n",
       "SGL_RS00635            Replication, recombination and repair     NaN  Op2588  \n",
       "SGL_RS01880  Translation, ribosomal structure and biogenesis     NaN   Op465  \n",
       "SGL_RS00015                                No COG annotation     NaN   Op686  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3690\n"
     ]
    }
   ],
   "source": [
    "#index is duplicated for uniprot records, so merge this cell values in one row\n",
    "problematic_colum = 'uniprot'\n",
    "duplicated_locus = set([ t for t in gene_table['locus_tag'] if gene_table['locus_tag'].tolist().count(t)>1])\n",
    "\n",
    "if len(duplicated_locus)>0:\n",
    "    replace_dict = {}\n",
    "    for locus in duplicated_locus:\n",
    "        group = gene_table.loc[gene_table['locus_tag']==locus]\n",
    "        new_value = ';'.join(group[problematic_colum].tolist())\n",
    "        for u_id in group[problematic_colum].tolist():\n",
    "            replace_dict[u_id] = new_value\n",
    "\n",
    "gene_table.replace({ problematic_colum : replace_dict},inplace=True)\n",
    "gene_table.drop_duplicates(inplace=True)\n",
    "gene_table = gene_table.set_index('locus_tag')\n",
    "\n",
    "display(gene_table.head())\n",
    "print(len(gene_table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "                   A = path.join(data_dir,'A.csv'),\n",
    "                   X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "                   gene_table = gene_table,\n",
    "                   sample_table = path.join(data_dir,'final_metadata.tsv'),\n",
    "#                  trn = path.join(data_dir,'TRN.csv'),\n",
    "                   threshold_method='kmeans')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you don't have a TRN (or have a very minimal TRN), use `threshold_method = 'kmeans'`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "#                    A = path.join(data_dir,'A.csv'),\n",
    "#                    X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "#                    gene_table = path.join(data_dir,'gene_info.csv'),\n",
    "#                    sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "#                    trn = path.join(data_dir,'TRN.csv'),\n",
    "#                    threshold_method = 'kmeans')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Regulatory iModulons\n",
    "Use `compute_trn_enrichment` to automatically check for Regulatory iModulons. The more complete your TRN, the more regulatory iModulons you'll find."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'gene_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_25178/464696660.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute_trn_enrichment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pymodulon/core.py\u001b[0m in \u001b[0;36mcompute_trn_enrichment\u001b[0;34m(self, imodulons, fdr, max_regs, save, method, force, evidence)\u001b[0m\n\u001b[1;32m    749\u001b[0m                 \u001b[0mfdr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfdr\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    750\u001b[0m                 \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 751\u001b[0;31m                 \u001b[0mforce\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mforce\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    752\u001b[0m             )\n\u001b[1;32m    753\u001b[0m             \u001b[0mdf_enriched\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"imodulon\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mimodulon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pymodulon/enrichment.py\u001b[0m in \u001b[0;36mcompute_trn_enrichment\u001b[0;34m(gene_set, all_genes, trn, max_regs, fdr, method, force)\u001b[0m\n\u001b[1;32m    248\u001b[0m     \u001b[0;31m# This reduces the total runtime by skipping unnecessary tests\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    249\u001b[0m     \u001b[0;31m# However, this needs to be taken into account for FDR\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 250\u001b[0;31m     \u001b[0mimod_regs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtrn\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtrn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgene_id\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_set\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregulator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    251\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    252\u001b[0m     \u001b[0;31m# # Account for imodulons with no regulators\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'gene_id'"
     ]
    }
   ],
   "source": [
    "ica_data.compute_trn_enrichment()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also search for AND/OR combinations of regulators using the `max_regs` argument.\n",
    "\n",
    "Regulator enrichments can be directly saved to the `imodulon_table` using the `save` argument. This saves the enrichment with the lowest q-value to the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'gene_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_25178/4002065078.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# First search for regulator enrichments with 2 regulators\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute_trn_enrichment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmax_regs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# Next, search for regulator enrichments with just one regulator. This will supercede the 2 regulator enrichments.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute_trn_enrichment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmax_regs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pymodulon/core.py\u001b[0m in \u001b[0;36mcompute_trn_enrichment\u001b[0;34m(self, imodulons, fdr, max_regs, save, method, force, evidence)\u001b[0m\n\u001b[1;32m    749\u001b[0m                 \u001b[0mfdr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfdr\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    750\u001b[0m                 \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 751\u001b[0;31m                 \u001b[0mforce\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mforce\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    752\u001b[0m             )\n\u001b[1;32m    753\u001b[0m             \u001b[0mdf_enriched\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"imodulon\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mimodulon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pymodulon/enrichment.py\u001b[0m in \u001b[0;36mcompute_trn_enrichment\u001b[0;34m(gene_set, all_genes, trn, max_regs, fdr, method, force)\u001b[0m\n\u001b[1;32m    248\u001b[0m     \u001b[0;31m# This reduces the total runtime by skipping unnecessary tests\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    249\u001b[0m     \u001b[0;31m# However, this needs to be taken into account for FDR\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 250\u001b[0;31m     \u001b[0mimod_regs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtrn\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtrn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgene_id\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_set\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregulator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    251\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    252\u001b[0m     \u001b[0;31m# # Account for imodulons with no regulators\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'gene_id'"
     ]
    }
   ],
   "source": [
    "# First search for regulator enrichments with 2 regulators\n",
    "ica_data.compute_trn_enrichment(max_regs=2,save=True)\n",
    "\n",
    "# Next, search for regulator enrichments with just one regulator. This will supercede the 2 regulator enrichments.\n",
    "ica_data.compute_trn_enrichment(max_regs=1,save=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The list of regulatory iModulons are shown below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'regulator'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_25178/2985153385.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mregulatory_imodulons\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregulator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnotnull\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'Total iModulons'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mregulatory_imodulons\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'Regulatory iModulons'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mregulatory_imodulons\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'regulator'"
     ]
    }
   ],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]\n",
    "print(len(ica_data.imodulon_table),'Total iModulons')\n",
    "print(len(regulatory_imodulons),'Regulatory iModulons')\n",
    "regulatory_imodulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can rename iModulons in this jupyter notebook, or you can save the iModulon table as a CSV and edit it in Excel.\n",
    "\n",
    "If two iModulons have the same regulator (e.g. 'Reg'), they will be named 'Reg-1' and 'Reg-2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'regulatory_imodulons' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_25178/2777212077.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrename_imodulons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mregulatory_imodulons\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregulator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'regulatory_imodulons' is not defined"
     ]
    }
   ],
   "source": [
    "ica_data.rename_imodulons(regulatory_imodulons.regulator.to_dict())\n",
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'regulator'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_25178/1171916079.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mregulatory_imodulons\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregulator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnotnull\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'regulator'"
     ]
    }
   ],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functional iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GO annotations and KEGG pathways/modules were generated in the 1_create_the_gene_table.ipynb notebook. Enrichments will be calculated in this notebook, and further curated in the 3_manual_iModulon_curation notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First load the Gene Ontology annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_ontology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sll0537</td>\n",
       "      <td>sll0537</td>\n",
       "      <td>ammonium homeostasis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sll0537</td>\n",
       "      <td>sll0537</td>\n",
       "      <td>ammonium transmembrane transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sll0601</td>\n",
       "      <td>sll0601</td>\n",
       "      <td>N-carbamoylputrescine amidase activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sll0601</td>\n",
       "      <td>sll0601</td>\n",
       "      <td>putrescine biosynthetic process from arginine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>slr1145</td>\n",
       "      <td>gltS</td>\n",
       "      <td>glutamate synthase (NADH) activity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id gene_name                                  gene_ontology\n",
       "3  sll0537   sll0537                           ammonium homeostasis\n",
       "4  sll0537   sll0537               ammonium transmembrane transport\n",
       "5  sll0601   sll0601         N-carbamoylputrescine amidase activity\n",
       "6  sll0601   sll0601  putrescine biosynthetic process from arginine\n",
       "7  slr1145      gltS             glutamate synthase (NADH) activity"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2254\n"
     ]
    }
   ],
   "source": [
    "DF_GO = pd.read_csv(path.join(external_data,'GO_annotations_curated.csv'),index_col=0)\n",
    "display(DF_GO.head())\n",
    "print(len(DF_GO))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>operon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SGL_RS01370</td>\n",
       "      <td>SGL_RS01370</td>\n",
       "      <td>NC_005232.1</td>\n",
       "      <td>slr6001</td>\n",
       "      <td>243</td>\n",
       "      <td>3053</td>\n",
       "      <td>+</td>\n",
       "      <td>hybrid sensor histidine kinase/response regulator</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SGL_RS00010</td>\n",
       "      <td>SGL_RS00010</td>\n",
       "      <td>NC_005229.1</td>\n",
       "      <td>sll5002</td>\n",
       "      <td>574</td>\n",
       "      <td>1017</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SGL_RS00635</td>\n",
       "      <td>SGL_RS00635</td>\n",
       "      <td>NC_005230.1</td>\n",
       "      <td>sll7002</td>\n",
       "      <td>706</td>\n",
       "      <td>1287</td>\n",
       "      <td>-</td>\n",
       "      <td>IS701 family transposase</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op2588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SGL_RS01880</td>\n",
       "      <td>SGL_RS01880</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0612</td>\n",
       "      <td>811</td>\n",
       "      <td>1494</td>\n",
       "      <td>+</td>\n",
       "      <td>pseudouridine synthase</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SGL_RS00015</td>\n",
       "      <td>SGL_RS00015</td>\n",
       "      <td>NC_005229.1</td>\n",
       "      <td>sll5003</td>\n",
       "      <td>1136</td>\n",
       "      <td>1540</td>\n",
       "      <td>-</td>\n",
       "      <td>DUF5615 family PIN-like protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3685</th>\n",
       "      <td>SGL_RS18370</td>\n",
       "      <td>SGL_RS18370</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0607</td>\n",
       "      <td>3569134</td>\n",
       "      <td>3569523</td>\n",
       "      <td>+</td>\n",
       "      <td>cyclic nucleotide-binding domain-containing pr...</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3686</th>\n",
       "      <td>SGL_RS18375</td>\n",
       "      <td>hisIE</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0608</td>\n",
       "      <td>3569672</td>\n",
       "      <td>3570319</td>\n",
       "      <td>+</td>\n",
       "      <td>bifunctional phosphoribosyl-AMP cyclohydrolase...</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>A0A6P1VM37_9SYNC</td>\n",
       "      <td>Op2513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3687</th>\n",
       "      <td>SGL_RS18380</td>\n",
       "      <td>SGL_RS18380</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0609</td>\n",
       "      <td>3570424</td>\n",
       "      <td>3571575</td>\n",
       "      <td>+</td>\n",
       "      <td>GTP-binding protein</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3688</th>\n",
       "      <td>SGL_RS18385</td>\n",
       "      <td>SGL_RS18385</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0610</td>\n",
       "      <td>3571711</td>\n",
       "      <td>3572403</td>\n",
       "      <td>+</td>\n",
       "      <td>ABC transporter permease</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Op108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689</th>\n",
       "      <td>SGL_RS01875</td>\n",
       "      <td>sds</td>\n",
       "      <td>NC_000911.1</td>\n",
       "      <td>slr0611</td>\n",
       "      <td>3573271</td>\n",
       "      <td>3574242</td>\n",
       "      <td>+</td>\n",
       "      <td>solanesyl diphosphate synthase</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>A0A6P1VKP0_9SYNC</td>\n",
       "      <td>Op1759</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3690 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        locus_tag    gene_name    accession old_locus_tag    start      end  \\\n",
       "0     SGL_RS01370  SGL_RS01370  NC_005232.1       slr6001      243     3053   \n",
       "1     SGL_RS00010  SGL_RS00010  NC_005229.1       sll5002      574     1017   \n",
       "2     SGL_RS00635  SGL_RS00635  NC_005230.1       sll7002      706     1287   \n",
       "3     SGL_RS01880  SGL_RS01880  NC_000911.1       slr0612      811     1494   \n",
       "4     SGL_RS00015  SGL_RS00015  NC_005229.1       sll5003     1136     1540   \n",
       "...           ...          ...          ...           ...      ...      ...   \n",
       "3685  SGL_RS18370  SGL_RS18370  NC_000911.1       slr0607  3569134  3569523   \n",
       "3686  SGL_RS18375        hisIE  NC_000911.1       slr0608  3569672  3570319   \n",
       "3687  SGL_RS18380  SGL_RS18380  NC_000911.1       slr0609  3570424  3571575   \n",
       "3688  SGL_RS18385  SGL_RS18385  NC_000911.1       slr0610  3571711  3572403   \n",
       "3689  SGL_RS01875          sds  NC_000911.1       slr0611  3573271  3574242   \n",
       "\n",
       "     strand                                       gene_product  \\\n",
       "0         +  hybrid sensor histidine kinase/response regulator   \n",
       "1         -                               hypothetical protein   \n",
       "2         -                           IS701 family transposase   \n",
       "3         +                             pseudouridine synthase   \n",
       "4         -                    DUF5615 family PIN-like protein   \n",
       "...     ...                                                ...   \n",
       "3685      +  cyclic nucleotide-binding domain-containing pr...   \n",
       "3686      +  bifunctional phosphoribosyl-AMP cyclohydrolase...   \n",
       "3687      +                                GTP-binding protein   \n",
       "3688      +                           ABC transporter permease   \n",
       "3689      +                     solanesyl diphosphate synthase   \n",
       "\n",
       "                                                  COG           uniprot  \\\n",
       "0                      Signal transduction mechanisms               NaN   \n",
       "1                                   No COG annotation               NaN   \n",
       "2               Replication, recombination and repair               NaN   \n",
       "3     Translation, ribosomal structure and biogenesis               NaN   \n",
       "4                                   No COG annotation               NaN   \n",
       "...                                               ...               ...   \n",
       "3685                   Signal transduction mechanisms               NaN   \n",
       "3686              Nucleotide transport and metabolism  A0A6P1VM37_9SYNC   \n",
       "3687                                 Function unknown               NaN   \n",
       "3688                                 Function unknown               NaN   \n",
       "3689                Coenzyme transport and metabolism  A0A6P1VKP0_9SYNC   \n",
       "\n",
       "      operon  \n",
       "0       Op97  \n",
       "1      Op712  \n",
       "2     Op2588  \n",
       "3      Op465  \n",
       "4      Op686  \n",
       "...      ...  \n",
       "3685   Op766  \n",
       "3686  Op2513  \n",
       "3687   Op489  \n",
       "3688   Op108  \n",
       "3689  Op1759  \n",
       "\n",
       "[3690 rows x 11 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table = gene_table.reset_index()\n",
    "gene_table = gene_table.rename(columns={'index':'locus_tag'})\n",
    "gene_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_ontology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SGL_RS16895</td>\n",
       "      <td>sll0537</td>\n",
       "      <td>ammonium homeostasis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SGL_RS16895</td>\n",
       "      <td>sll0537</td>\n",
       "      <td>ammonium transmembrane transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SGL_RS14360</td>\n",
       "      <td>sll0601</td>\n",
       "      <td>N-carbamoylputrescine amidase activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SGL_RS14360</td>\n",
       "      <td>sll0601</td>\n",
       "      <td>putrescine biosynthetic process from arginine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SGL_RS05620</td>\n",
       "      <td>gltS</td>\n",
       "      <td>glutamate synthase (NADH) activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3351</th>\n",
       "      <td>SGL_RS02565</td>\n",
       "      <td>sll0225</td>\n",
       "      <td>cytoplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3354</th>\n",
       "      <td>SGL_RS02545</td>\n",
       "      <td>slr0245</td>\n",
       "      <td>transcription factor binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3355</th>\n",
       "      <td>SGL_RS02545</td>\n",
       "      <td>slr0245</td>\n",
       "      <td>negative regulation of transcription by RNA po...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3356</th>\n",
       "      <td>SGL_RS02545</td>\n",
       "      <td>slr0245</td>\n",
       "      <td>histone deacetylase activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3357</th>\n",
       "      <td>SGL_RS02545</td>\n",
       "      <td>slr0245</td>\n",
       "      <td>histone deacetylation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2254 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_id gene_name                                      gene_ontology\n",
       "3     SGL_RS16895   sll0537                               ammonium homeostasis\n",
       "4     SGL_RS16895   sll0537                   ammonium transmembrane transport\n",
       "5     SGL_RS14360   sll0601             N-carbamoylputrescine amidase activity\n",
       "6     SGL_RS14360   sll0601      putrescine biosynthetic process from arginine\n",
       "7     SGL_RS05620      gltS                 glutamate synthase (NADH) activity\n",
       "...           ...       ...                                                ...\n",
       "3351  SGL_RS02565   sll0225                                          cytoplasm\n",
       "3354  SGL_RS02545   slr0245                       transcription factor binding\n",
       "3355  SGL_RS02545   slr0245  negative regulation of transcription by RNA po...\n",
       "3356  SGL_RS02545   slr0245                       histone deacetylase activity\n",
       "3357  SGL_RS02545   slr0245                              histone deacetylation\n",
       "\n",
       "[2254 rows x 3 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_features = ['old_locus_tag','locus_tag']\n",
    "mapping_df = gene_table.loc[gene_table.old_locus_tag.isin(DF_GO.gene_id.tolist())]\n",
    "go_mapping_dict = mapping_df[mapping_features].set_index(mapping_features[0]).to_dict()[mapping_features[1]]\n",
    "DF_GO.replace({'gene_id' : go_mapping_dict}, inplace=True)\n",
    "DF_GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich = ica_data.compute_annotation_enrichment(DF_GO,'gene_ontology')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>gene_ontology</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>structural constituent of ribosome</td>\n",
       "      <td>4.033821e-29</td>\n",
       "      <td>3.513458e-26</td>\n",
       "      <td>0.313433</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.407767</td>\n",
       "      <td>21.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>cytosolic large ribosomal subunit</td>\n",
       "      <td>1.095099e-21</td>\n",
       "      <td>4.769155e-19</td>\n",
       "      <td>0.208955</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.325581</td>\n",
       "      <td>14.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>cytosolic small ribosomal subunit</td>\n",
       "      <td>3.988856e-07</td>\n",
       "      <td>1.158098e-04</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.129870</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>translation</td>\n",
       "      <td>1.218767e-06</td>\n",
       "      <td>2.653865e-04</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>0.126582</td>\n",
       "      <td>5.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>RNA binding</td>\n",
       "      <td>5.577349e-05</td>\n",
       "      <td>9.715741e-03</td>\n",
       "      <td>0.044776</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>cytoplasmic translation</td>\n",
       "      <td>3.248506e-04</td>\n",
       "      <td>4.042070e-02</td>\n",
       "      <td>0.029851</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.057971</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>large ribosomal subunit rRNA binding</td>\n",
       "      <td>3.248506e-04</td>\n",
       "      <td>4.042070e-02</td>\n",
       "      <td>0.029851</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.057971</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>17</td>\n",
       "      <td>plasma membrane-derived thylakoid photosystem II</td>\n",
       "      <td>6.278898e-05</td>\n",
       "      <td>5.468920e-02</td>\n",
       "      <td>0.059701</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon                                     gene_ontology        pvalue  \\\n",
       "0         1                structural constituent of ribosome  4.033821e-29   \n",
       "1         1                 cytosolic large ribosomal subunit  1.095099e-21   \n",
       "2         1                 cytosolic small ribosomal subunit  3.988856e-07   \n",
       "3         1                                       translation  1.218767e-06   \n",
       "4         1                                       RNA binding  5.577349e-05   \n",
       "5         1                           cytoplasmic translation  3.248506e-04   \n",
       "6         1              large ribosomal subunit rRNA binding  3.248506e-04   \n",
       "7        17  plasma membrane-derived thylakoid photosystem II  6.278898e-05   \n",
       "\n",
       "         qvalue  precision    recall   f1score    TP  target_set_size  \\\n",
       "0  3.513458e-26   0.313433  0.583333  0.407767  21.0             36.0   \n",
       "1  4.769155e-19   0.208955  0.736842  0.325581  14.0             19.0   \n",
       "2  1.158098e-04   0.074627  0.500000  0.129870   5.0             10.0   \n",
       "3  2.653865e-04   0.074627  0.416667  0.126582   5.0             12.0   \n",
       "4  9.715741e-03   0.044776  0.600000  0.083333   3.0              5.0   \n",
       "5  4.042070e-02   0.029851  1.000000  0.057971   2.0              2.0   \n",
       "6  4.042070e-02   0.029851  1.000000  0.057971   2.0              2.0   \n",
       "7  5.468920e-02   0.059701  0.307692  0.100000   4.0             16.0   \n",
       "\n",
       "   imodulon_size  \n",
       "0           67.0  \n",
       "1           67.0  \n",
       "2           67.0  \n",
       "3           67.0  \n",
       "4           67.0  \n",
       "5           67.0  \n",
       "6           67.0  \n",
       "7           67.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    }
   ],
   "source": [
    "display(DF_GO_enrich)\n",
    "print(len(DF_GO_enrich))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SGL_RS05715',\n",
       " 'SGL_RS05720',\n",
       " 'SGL_RS05730',\n",
       " 'SGL_RS05765',\n",
       " 'SGL_RS05770',\n",
       " 'SGL_RS05780',\n",
       " 'SGL_RS05785',\n",
       " 'SGL_RS05790',\n",
       " 'SGL_RS05800',\n",
       " 'SGL_RS05805',\n",
       " 'SGL_RS05815',\n",
       " 'SGL_RS05820',\n",
       " 'SGL_RS05825',\n",
       " 'SGL_RS05830',\n",
       " 'SGL_RS05835',\n",
       " 'SGL_RS05845',\n",
       " 'SGL_RS05850',\n",
       " 'SGL_RS06195',\n",
       " 'SGL_RS06200',\n",
       " 'SGL_RS06210',\n",
       " 'SGL_RS10920'}"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(ica_data.view_imodulon(1).index).intersection(set(DF_GO.loc[DF_GO['gene_ontology']=='structural constituent of ribosome'].gene_id.tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SGL_RS08200', 'SGL_RS11935', 'SGL_RS14825', 'SGL_RS16825'}"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(ica_data.view_imodulon(17).index).intersection(set(DF_GO.loc[DF_GO['gene_ontology']=='plasma membrane-derived thylakoid photosystem II'].gene_id.tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "856 genes are annotated\n",
      "0 2\n",
      "1 21\n",
      "2 9\n",
      "3 2\n",
      "4 2\n",
      "5 11\n",
      "6 nan\n",
      "7 8\n",
      "8 1\n",
      "9 8\n",
      "10 nan\n",
      "11 39\n",
      "12 2\n",
      "13 6\n",
      "14 6\n",
      "15 9\n",
      "16 3\n",
      "17 7\n",
      "18 1\n",
      "19 6\n",
      "265 / 1309 of imodulon genes are present in GO annotation table\n"
     ]
    }
   ],
   "source": [
    "imodulon_genes = set()\n",
    "no_annotation = ['No COG annotation', 'Function unknown']\n",
    "annotated_genes = DF_GO.loc[~DF_GO['gene_ontology'].isin(no_annotation)]\n",
    "print('%s genes are annotated' % len(annotated_genes.gene_id.unique()))\n",
    "for i in ica_data.imodulon_table.index:\n",
    "    print( i , annotated_genes.loc[annotated_genes.gene_id.isin(ica_data.view_imodulon(i).index.tolist()), 'gene_ontology'].value_counts().max())\n",
    "    imodulon_genes |= set(ica_data.view_imodulon(i).index.tolist())\n",
    "\n",
    "print('%s / %s of imodulon genes are present in GO annotation table' % (len(imodulon_genes.intersection(set(annotated_genes.gene_id.tolist()))), len(imodulon_genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "plasma membrane-derived thylakoid photosystem II                  7\n",
       "protein binding                                                   7\n",
       "plasma membrane-derived photosystem I                             2\n",
       "ammonium homeostasis                                              1\n",
       "translation                                                       1\n",
       "cytoplasm                                                         1\n",
       "glutamine biosynthetic process                                    1\n",
       "membrane                                                          1\n",
       "glutamate-ammonia ligase activity                                 1\n",
       "water channel activity                                            1\n",
       "plasma membrane                                                   1\n",
       "mRNA binding                                                      1\n",
       "cytosolic large ribosomal subunit                                 1\n",
       "structural constituent of ribosome                                1\n",
       "ammonium transmembrane transport                                  1\n",
       "proton motive force-driven ATP synthesis                          1\n",
       "proton-transporting ATP synthase complex, coupling factor F(o)    1\n",
       "photosystem II                                                    1\n",
       "ubiquinol-cytochrome-c reductase activity                         1\n",
       "cytochrome b6f complex                                            1\n",
       "phycobilisome                                                     1\n",
       "nitrogen utilization                                              1\n",
       "Name: gene_ontology, dtype: int64"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated_genes.loc[annotated_genes.gene_id.isin(ica_data.view_imodulon(17).index.tolist()), 'gene_ontology'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KEGG Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load KEGG mapping\n",
    "The `kegg_mapping.csv` file contains KEGG orthologies, pathways, modules, and reactions. Only pathways and modules are relevant to iModulon characterization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KEGG_orth' 'KEGG_pathway' 'KEGG_module' 'KEGG_reaction']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>database</th>\n",
       "      <th>kegg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SGL_RS01885</td>\n",
       "      <td>KEGG_orth</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SGL_RS01890</td>\n",
       "      <td>KEGG_orth</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SGL_RS01925</td>\n",
       "      <td>KEGG_orth</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SGL_RS01955</td>\n",
       "      <td>KEGG_orth</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SGL_RS01985</td>\n",
       "      <td>KEGG_orth</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_id   database kegg_id\n",
       "1   SGL_RS01885  KEGG_orth       -\n",
       "2   SGL_RS01890  KEGG_orth       -\n",
       "9   SGL_RS01925  KEGG_orth       -\n",
       "15  SGL_RS01955  KEGG_orth       -\n",
       "17  SGL_RS01985  KEGG_orth       -"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_KEGG = pd.read_csv(path.join(external_data,'kegg_mapping.csv'),index_col=0)\n",
    "print(DF_KEGG.database.unique())\n",
    "DF_KEGG.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways = DF_KEGG[DF_KEGG.database == 'KEGG_pathway']\n",
    "kegg_modules = DF_KEGG[DF_KEGG.database == 'KEGG_module']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform enrichment\n",
    "Uses the `compute_annotation_enrichment` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_pathway_enrich = ica_data.compute_annotation_enrichment(kegg_pathways,'kegg_id')\n",
    "DF_module_enrich = ica_data.compute_annotation_enrichment(kegg_modules,'kegg_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>map04141</td>\n",
       "      <td>2.962003e-04</td>\n",
       "      <td>6.249826e-02</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>64.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>map03010</td>\n",
       "      <td>1.563306e-33</td>\n",
       "      <td>3.298575e-31</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.490566</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>26.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>map02010</td>\n",
       "      <td>5.167328e-09</td>\n",
       "      <td>1.090306e-06</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.160714</td>\n",
       "      <td>9.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>map02010</td>\n",
       "      <td>6.517585e-06</td>\n",
       "      <td>1.375210e-03</td>\n",
       "      <td>0.212121</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.119658</td>\n",
       "      <td>7.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>map00196</td>\n",
       "      <td>2.623274e-05</td>\n",
       "      <td>5.535109e-03</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.090090</td>\n",
       "      <td>5.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0         0  map04141  2.962003e-04  6.249826e-02   0.031250  1.000000   \n",
       "1         1  map03010  1.563306e-33  3.298575e-31   0.388060  0.490566   \n",
       "2         3  map02010  5.167328e-09  1.090306e-06   0.321429  0.107143   \n",
       "3         9  map02010  6.517585e-06  1.375210e-03   0.212121  0.083333   \n",
       "4        13  map00196  2.623274e-05  5.535109e-03   0.052083  0.333333   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \n",
       "0  0.060606   2.0              2.0           64.0  \n",
       "1  0.433333  26.0             53.0           67.0  \n",
       "2  0.160714   9.0             84.0           28.0  \n",
       "3  0.119658   7.0             84.0           33.0  \n",
       "4  0.090090   5.0             15.0           96.0  "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M00178</td>\n",
       "      <td>1.563306e-33</td>\n",
       "      <td>3.361107e-31</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.490566</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>26.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>M00179</td>\n",
       "      <td>1.114389e-31</td>\n",
       "      <td>1.197968e-29</td>\n",
       "      <td>0.313433</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>0.432990</td>\n",
       "      <td>21.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>M00222</td>\n",
       "      <td>1.087344e-18</td>\n",
       "      <td>2.337790e-16</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.818182</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>M00316</td>\n",
       "      <td>4.350375e-08</td>\n",
       "      <td>9.353306e-06</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>M00246</td>\n",
       "      <td>5.352880e-05</td>\n",
       "      <td>1.150869e-02</td>\n",
       "      <td>0.012618</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.024922</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>317.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0         1  M00178  1.563306e-33  3.361107e-31   0.388060  0.490566   \n",
       "1         1  M00179  1.114389e-31  1.197968e-29   0.313433  0.700000   \n",
       "2         3  M00222  1.087344e-18  2.337790e-16   0.321429  0.818182   \n",
       "3         4  M00316  4.350375e-08  9.353306e-06   0.214286  1.000000   \n",
       "4         5  M00246  5.352880e-05  1.150869e-02   0.012618  1.000000   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \n",
       "0  0.433333  26.0             53.0           67.0  \n",
       "1  0.432990  21.0             30.0           67.0  \n",
       "2  0.461538   9.0             11.0           28.0  \n",
       "3  0.352941   3.0              3.0           14.0  \n",
       "4  0.024922   4.0              4.0          317.0  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert KEGG IDs to human-readable names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7cc2915e267d438c8a18e05508242f49",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/15 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d60f3754a1c048f58bd9be461bfbf586",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/17 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for idx,key in tqdm(DF_pathway_enrich.kegg_id.items(),total=len(DF_pathway_enrich)):\n",
    "    text = REST.kegg_find('pathway',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = None\n",
    "    \n",
    "for idx,key in tqdm(DF_module_enrich.kegg_id.items(),total=len(DF_module_enrich)):\n",
    "    text = REST.kegg_find('module',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_module_enrich.loc[idx,'module_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_module_enrich.loc[idx,'module_name'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>map04141</td>\n",
       "      <td>2.962003e-04</td>\n",
       "      <td>6.249826e-02</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>map03010</td>\n",
       "      <td>1.563306e-33</td>\n",
       "      <td>3.298575e-31</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.490566</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>26.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Ribosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>map02010</td>\n",
       "      <td>5.167328e-09</td>\n",
       "      <td>1.090306e-06</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.160714</td>\n",
       "      <td>9.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>ABC transporters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>map02010</td>\n",
       "      <td>6.517585e-06</td>\n",
       "      <td>1.375210e-03</td>\n",
       "      <td>0.212121</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.119658</td>\n",
       "      <td>7.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>ABC transporters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>map00196</td>\n",
       "      <td>2.623274e-05</td>\n",
       "      <td>5.535109e-03</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.090090</td>\n",
       "      <td>5.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Photosynthesis - antenna proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>14</td>\n",
       "      <td>map00910</td>\n",
       "      <td>3.164213e-10</td>\n",
       "      <td>6.676489e-08</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.292683</td>\n",
       "      <td>6.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Nitrogen metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>14</td>\n",
       "      <td>map02010</td>\n",
       "      <td>2.448962e-04</td>\n",
       "      <td>1.498222e-02</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>0.059524</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>5.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>ABC transporters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>14</td>\n",
       "      <td>map04724</td>\n",
       "      <td>1.426302e-04</td>\n",
       "      <td>1.498222e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.137931</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Glutamatergic synapse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>14</td>\n",
       "      <td>map04727</td>\n",
       "      <td>2.840232e-04</td>\n",
       "      <td>1.498222e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>GABAergic synapse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>14</td>\n",
       "      <td>map04217</td>\n",
       "      <td>4.713197e-04</td>\n",
       "      <td>1.988969e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Necroptosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>15</td>\n",
       "      <td>map00195</td>\n",
       "      <td>1.372258e-04</td>\n",
       "      <td>2.895464e-02</td>\n",
       "      <td>0.053571</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>0.085106</td>\n",
       "      <td>12.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>224.0</td>\n",
       "      <td>Photosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>17</td>\n",
       "      <td>map00195</td>\n",
       "      <td>1.855171e-23</td>\n",
       "      <td>3.914412e-21</td>\n",
       "      <td>0.313433</td>\n",
       "      <td>0.362069</td>\n",
       "      <td>0.336000</td>\n",
       "      <td>21.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Photosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>17</td>\n",
       "      <td>map00196</td>\n",
       "      <td>5.444976e-15</td>\n",
       "      <td>5.744450e-13</td>\n",
       "      <td>0.149254</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.243902</td>\n",
       "      <td>10.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Photosynthesis - antenna proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>17</td>\n",
       "      <td>map01100</td>\n",
       "      <td>6.931914e-12</td>\n",
       "      <td>4.875446e-10</td>\n",
       "      <td>0.507463</td>\n",
       "      <td>0.060391</td>\n",
       "      <td>0.107937</td>\n",
       "      <td>34.0</td>\n",
       "      <td>563.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Metabolic pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>17</td>\n",
       "      <td>map00710</td>\n",
       "      <td>2.509130e-04</td>\n",
       "      <td>1.323566e-02</td>\n",
       "      <td>0.059701</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.094118</td>\n",
       "      <td>4.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Carbon fixation in photosynthetic organisms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0          0  map04141  2.962003e-04  6.249826e-02   0.031250  1.000000   \n",
       "1          1  map03010  1.563306e-33  3.298575e-31   0.388060  0.490566   \n",
       "2          3  map02010  5.167328e-09  1.090306e-06   0.321429  0.107143   \n",
       "3          9  map02010  6.517585e-06  1.375210e-03   0.212121  0.083333   \n",
       "4         13  map00196  2.623274e-05  5.535109e-03   0.052083  0.333333   \n",
       "5         14  map00910  3.164213e-10  6.676489e-08   0.230769  0.400000   \n",
       "6         14  map02010  2.448962e-04  1.498222e-02   0.192308  0.059524   \n",
       "7         14  map04724  1.426302e-04  1.498222e-02   0.076923  0.666667   \n",
       "8         14  map04727  2.840232e-04  1.498222e-02   0.076923  0.500000   \n",
       "9         14  map04217  4.713197e-04  1.988969e-02   0.076923  0.400000   \n",
       "10        15  map00195  1.372258e-04  2.895464e-02   0.053571  0.206897   \n",
       "11        17  map00195  1.855171e-23  3.914412e-21   0.313433  0.362069   \n",
       "12        17  map00196  5.444976e-15  5.744450e-13   0.149254  0.666667   \n",
       "13        17  map01100  6.931914e-12  4.875446e-10   0.507463  0.060391   \n",
       "14        17  map00710  2.509130e-04  1.323566e-02   0.059701  0.222222   \n",
       "\n",
       "     f1score    TP  target_set_size  imodulon_size  \\\n",
       "0   0.060606   2.0              2.0           64.0   \n",
       "1   0.433333  26.0             53.0           67.0   \n",
       "2   0.160714   9.0             84.0           28.0   \n",
       "3   0.119658   7.0             84.0           33.0   \n",
       "4   0.090090   5.0             15.0           96.0   \n",
       "5   0.292683   6.0             15.0           26.0   \n",
       "6   0.090909   5.0             84.0           26.0   \n",
       "7   0.137931   2.0              3.0           26.0   \n",
       "8   0.133333   2.0              4.0           26.0   \n",
       "9   0.129032   2.0              5.0           26.0   \n",
       "10  0.085106  12.0             58.0          224.0   \n",
       "11  0.336000  21.0             58.0           67.0   \n",
       "12  0.243902  10.0             15.0           67.0   \n",
       "13  0.107937  34.0            563.0           67.0   \n",
       "14  0.094118   4.0             18.0           67.0   \n",
       "\n",
       "                                   pathway_name  \n",
       "0   Protein processing in endoplasmic reticulum  \n",
       "1                                      Ribosome  \n",
       "2                              ABC transporters  \n",
       "3                              ABC transporters  \n",
       "4             Photosynthesis - antenna proteins  \n",
       "5                           Nitrogen metabolism  \n",
       "6                              ABC transporters  \n",
       "7                         Glutamatergic synapse  \n",
       "8                             GABAergic synapse  \n",
       "9                                   Necroptosis  \n",
       "10                               Photosynthesis  \n",
       "11                               Photosynthesis  \n",
       "12            Photosynthesis - antenna proteins  \n",
       "13                           Metabolic pathways  \n",
       "14  Carbon fixation in photosynthetic organisms  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>module_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M00178</td>\n",
       "      <td>1.563306e-33</td>\n",
       "      <td>3.361107e-31</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.490566</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>26.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>M00179</td>\n",
       "      <td>1.114389e-31</td>\n",
       "      <td>1.197968e-29</td>\n",
       "      <td>0.313433</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>0.432990</td>\n",
       "      <td>21.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>M00222</td>\n",
       "      <td>1.087344e-18</td>\n",
       "      <td>2.337790e-16</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.818182</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>M00316</td>\n",
       "      <td>4.350375e-08</td>\n",
       "      <td>9.353306e-06</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>M00246</td>\n",
       "      <td>5.352880e-05</td>\n",
       "      <td>1.150869e-02</td>\n",
       "      <td>0.012618</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.024922</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>317.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>M00245</td>\n",
       "      <td>6.285372e-04</td>\n",
       "      <td>6.756775e-02</td>\n",
       "      <td>0.009464</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.018750</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>317.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9</td>\n",
       "      <td>M00240</td>\n",
       "      <td>2.636184e-08</td>\n",
       "      <td>5.667795e-06</td>\n",
       "      <td>0.121212</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9</td>\n",
       "      <td>M00190</td>\n",
       "      <td>6.441446e-06</td>\n",
       "      <td>6.924554e-04</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>13</td>\n",
       "      <td>M00155</td>\n",
       "      <td>3.225298e-04</td>\n",
       "      <td>6.934391e-02</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Cytochrome c oxidase, prokaryotes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13</td>\n",
       "      <td>M00154</td>\n",
       "      <td>6.699769e-04</td>\n",
       "      <td>7.202251e-02</td>\n",
       "      <td>0.020833</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.040816</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Cytochrome c oxidase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>13</td>\n",
       "      <td>M00004</td>\n",
       "      <td>1.279511e-03</td>\n",
       "      <td>9.169826e-02</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Pentose phosphate pathway (Pentose phosphate c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>14</td>\n",
       "      <td>M00323</td>\n",
       "      <td>3.078404e-06</td>\n",
       "      <td>6.618568e-04</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.193548</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>14</td>\n",
       "      <td>M00438</td>\n",
       "      <td>2.840232e-04</td>\n",
       "      <td>3.053249e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>17</td>\n",
       "      <td>M00161</td>\n",
       "      <td>2.299900e-09</td>\n",
       "      <td>4.944786e-07</td>\n",
       "      <td>0.089552</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Photosystem II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>17</td>\n",
       "      <td>M00163</td>\n",
       "      <td>1.004887e-08</td>\n",
       "      <td>1.080253e-06</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.136986</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Photosystem I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>17</td>\n",
       "      <td>M00165</td>\n",
       "      <td>6.278898e-05</td>\n",
       "      <td>4.499877e-03</td>\n",
       "      <td>0.059701</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>19</td>\n",
       "      <td>M00316</td>\n",
       "      <td>1.741267e-04</td>\n",
       "      <td>3.743725e-02</td>\n",
       "      <td>0.014493</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    imodulon kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0          1  M00178  1.563306e-33  3.361107e-31   0.388060  0.490566   \n",
       "1          1  M00179  1.114389e-31  1.197968e-29   0.313433  0.700000   \n",
       "2          3  M00222  1.087344e-18  2.337790e-16   0.321429  0.818182   \n",
       "3          4  M00316  4.350375e-08  9.353306e-06   0.214286  1.000000   \n",
       "4          5  M00246  5.352880e-05  1.150869e-02   0.012618  1.000000   \n",
       "5          5  M00245  6.285372e-04  6.756775e-02   0.009464  1.000000   \n",
       "6          9  M00240  2.636184e-08  5.667795e-06   0.121212  0.800000   \n",
       "7          9  M00190  6.441446e-06  6.924554e-04   0.090909  0.600000   \n",
       "8         13  M00155  3.225298e-04  6.934391e-02   0.031250  0.500000   \n",
       "9         13  M00154  6.699769e-04  7.202251e-02   0.020833  1.000000   \n",
       "10        13  M00004  1.279511e-03  9.169826e-02   0.031250  0.333333   \n",
       "11        14  M00323  3.078404e-06  6.618568e-04   0.115385  0.600000   \n",
       "12        14  M00438  2.840232e-04  3.053249e-02   0.076923  0.500000   \n",
       "13        17  M00161  2.299900e-09  4.944786e-07   0.089552  0.666667   \n",
       "14        17  M00163  1.004887e-08  1.080253e-06   0.074627  0.833333   \n",
       "15        17  M00165  6.278898e-05  4.499877e-03   0.059701  0.307692   \n",
       "16        19  M00316  1.741267e-04  3.743725e-02   0.014493  1.000000   \n",
       "\n",
       "     f1score    TP  target_set_size  imodulon_size  \\\n",
       "0   0.433333  26.0             53.0           67.0   \n",
       "1   0.432990  21.0             30.0           67.0   \n",
       "2   0.461538   9.0             11.0           28.0   \n",
       "3   0.352941   3.0              3.0           14.0   \n",
       "4   0.024922   4.0              4.0          317.0   \n",
       "5   0.018750   3.0              3.0          317.0   \n",
       "6   0.210526   4.0              5.0           33.0   \n",
       "7   0.157895   3.0              5.0           33.0   \n",
       "8   0.058824   3.0              6.0           96.0   \n",
       "9   0.040816   2.0              2.0           96.0   \n",
       "10  0.057143   3.0              9.0           96.0   \n",
       "11  0.193548   3.0              5.0           26.0   \n",
       "12  0.133333   2.0              4.0           26.0   \n",
       "13  0.157895   6.0              9.0           67.0   \n",
       "14  0.136986   5.0              6.0           67.0   \n",
       "15  0.100000   4.0             13.0           67.0   \n",
       "16  0.028571   3.0              3.0          207.0   \n",
       "\n",
       "                                          module_name  \n",
       "0                                                 NaN  \n",
       "1                                                 NaN  \n",
       "2                                                 NaN  \n",
       "3                                                 NaN  \n",
       "4                                                 NaN  \n",
       "5                                                 NaN  \n",
       "6                                                 NaN  \n",
       "7                                                 NaN  \n",
       "8                   Cytochrome c oxidase, prokaryotes  \n",
       "9                                Cytochrome c oxidase  \n",
       "10  Pentose phosphate pathway (Pentose phosphate c...  \n",
       "11                                               None  \n",
       "12                                               None  \n",
       "13                                     Photosystem II  \n",
       "14                                      Photosystem I  \n",
       "15   Reductive pentose phosphate cycle (Calvin cycle)  \n",
       "16                                               None  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>module_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M00178</td>\n",
       "      <td>1.563306e-33</td>\n",
       "      <td>3.361107e-31</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.490566</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>26.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Ribosome, bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>M00179</td>\n",
       "      <td>1.114389e-31</td>\n",
       "      <td>1.197968e-29</td>\n",
       "      <td>0.313433</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>0.432990</td>\n",
       "      <td>21.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Ribosome, archaea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>M00222</td>\n",
       "      <td>1.087344e-18</td>\n",
       "      <td>2.337790e-16</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.818182</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>Phosphate transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>M00316</td>\n",
       "      <td>4.350375e-08</td>\n",
       "      <td>9.353306e-06</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Manganese transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>M00246</td>\n",
       "      <td>5.352880e-05</td>\n",
       "      <td>1.150869e-02</td>\n",
       "      <td>0.012618</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.024922</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>317.0</td>\n",
       "      <td>Nickel transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>M00245</td>\n",
       "      <td>6.285372e-04</td>\n",
       "      <td>6.756775e-02</td>\n",
       "      <td>0.009464</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.018750</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>317.0</td>\n",
       "      <td>Cobalt/nickel transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9</td>\n",
       "      <td>M00240</td>\n",
       "      <td>2.636184e-08</td>\n",
       "      <td>5.667795e-06</td>\n",
       "      <td>0.121212</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>Iron complex transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9</td>\n",
       "      <td>M00190</td>\n",
       "      <td>6.441446e-06</td>\n",
       "      <td>6.924554e-04</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>Iron(III) transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>13</td>\n",
       "      <td>M00155</td>\n",
       "      <td>3.225298e-04</td>\n",
       "      <td>6.934391e-02</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Cytochrome c oxidase, prokaryotes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13</td>\n",
       "      <td>M00154</td>\n",
       "      <td>6.699769e-04</td>\n",
       "      <td>7.202251e-02</td>\n",
       "      <td>0.020833</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.040816</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Cytochrome c oxidase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>13</td>\n",
       "      <td>M00004</td>\n",
       "      <td>1.279511e-03</td>\n",
       "      <td>9.169826e-02</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Pentose phosphate pathway (Pentose phosphate c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>14</td>\n",
       "      <td>M00323</td>\n",
       "      <td>3.078404e-06</td>\n",
       "      <td>6.618568e-04</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.193548</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Urea transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>14</td>\n",
       "      <td>M00438</td>\n",
       "      <td>2.840232e-04</td>\n",
       "      <td>3.053249e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Nitrate/nitrite transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>17</td>\n",
       "      <td>M00161</td>\n",
       "      <td>2.299900e-09</td>\n",
       "      <td>4.944786e-07</td>\n",
       "      <td>0.089552</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Photosystem II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>17</td>\n",
       "      <td>M00163</td>\n",
       "      <td>1.004887e-08</td>\n",
       "      <td>1.080253e-06</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.136986</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Photosystem I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>17</td>\n",
       "      <td>M00165</td>\n",
       "      <td>6.278898e-05</td>\n",
       "      <td>4.499877e-03</td>\n",
       "      <td>0.059701</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>19</td>\n",
       "      <td>M00316</td>\n",
       "      <td>1.741267e-04</td>\n",
       "      <td>3.743725e-02</td>\n",
       "      <td>0.014493</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>Manganese transport system</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    imodulon kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0          1  M00178  1.563306e-33  3.361107e-31   0.388060  0.490566   \n",
       "1          1  M00179  1.114389e-31  1.197968e-29   0.313433  0.700000   \n",
       "2          3  M00222  1.087344e-18  2.337790e-16   0.321429  0.818182   \n",
       "3          4  M00316  4.350375e-08  9.353306e-06   0.214286  1.000000   \n",
       "4          5  M00246  5.352880e-05  1.150869e-02   0.012618  1.000000   \n",
       "5          5  M00245  6.285372e-04  6.756775e-02   0.009464  1.000000   \n",
       "6          9  M00240  2.636184e-08  5.667795e-06   0.121212  0.800000   \n",
       "7          9  M00190  6.441446e-06  6.924554e-04   0.090909  0.600000   \n",
       "8         13  M00155  3.225298e-04  6.934391e-02   0.031250  0.500000   \n",
       "9         13  M00154  6.699769e-04  7.202251e-02   0.020833  1.000000   \n",
       "10        13  M00004  1.279511e-03  9.169826e-02   0.031250  0.333333   \n",
       "11        14  M00323  3.078404e-06  6.618568e-04   0.115385  0.600000   \n",
       "12        14  M00438  2.840232e-04  3.053249e-02   0.076923  0.500000   \n",
       "13        17  M00161  2.299900e-09  4.944786e-07   0.089552  0.666667   \n",
       "14        17  M00163  1.004887e-08  1.080253e-06   0.074627  0.833333   \n",
       "15        17  M00165  6.278898e-05  4.499877e-03   0.059701  0.307692   \n",
       "16        19  M00316  1.741267e-04  3.743725e-02   0.014493  1.000000   \n",
       "\n",
       "     f1score    TP  target_set_size  imodulon_size  \\\n",
       "0   0.433333  26.0             53.0           67.0   \n",
       "1   0.432990  21.0             30.0           67.0   \n",
       "2   0.461538   9.0             11.0           28.0   \n",
       "3   0.352941   3.0              3.0           14.0   \n",
       "4   0.024922   4.0              4.0          317.0   \n",
       "5   0.018750   3.0              3.0          317.0   \n",
       "6   0.210526   4.0              5.0           33.0   \n",
       "7   0.157895   3.0              5.0           33.0   \n",
       "8   0.058824   3.0              6.0           96.0   \n",
       "9   0.040816   2.0              2.0           96.0   \n",
       "10  0.057143   3.0              9.0           96.0   \n",
       "11  0.193548   3.0              5.0           26.0   \n",
       "12  0.133333   2.0              4.0           26.0   \n",
       "13  0.157895   6.0              9.0           67.0   \n",
       "14  0.136986   5.0              6.0           67.0   \n",
       "15  0.100000   4.0             13.0           67.0   \n",
       "16  0.028571   3.0              3.0          207.0   \n",
       "\n",
       "                                          module_name  \n",
       "0                                  Ribosome, bacteria  \n",
       "1                                   Ribosome, archaea  \n",
       "2                          Phosphate transport system  \n",
       "3                          Manganese transport system  \n",
       "4                             Nickel transport system  \n",
       "5                      Cobalt/nickel transport system  \n",
       "6                       Iron complex transport system  \n",
       "7                          Iron(III) transport system  \n",
       "8                   Cytochrome c oxidase, prokaryotes  \n",
       "9                                Cytochrome c oxidase  \n",
       "10  Pentose phosphate pathway (Pentose phosphate c...  \n",
       "11                              Urea transport system  \n",
       "12                   Nitrate/nitrite transport system  \n",
       "13                                     Photosystem II  \n",
       "14                                      Photosystem I  \n",
       "15   Reductive pentose phosphate cycle (Calvin cycle)  \n",
       "16                         Manganese transport system  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "module_dict = {'M00178' : 'Ribosome, bacteria',\n",
    "               'M00179' : 'Ribosome, archaea',\n",
    "               'M00222' : 'Phosphate transport system',\n",
    "               'M00316' : 'Manganese transport system',\n",
    "               'M00246' : 'Nickel transport system',\n",
    "               'M00245' : 'Cobalt/nickel transport system',\n",
    "               'M00240' : 'Iron complex transport system',\n",
    "               'M00190' : 'Iron(III) transport system',\n",
    "               'M00323' : 'Urea transport system',\n",
    "               'M00438' : 'Nitrate/nitrite transport system',\n",
    "              }\n",
    "\n",
    "DF_module_enrich['module_name'] = [ module_dict[row['kegg_id']] if row['kegg_id'] in module_dict.keys() else row['module_name']\n",
    "                                    for index, row in DF_module_enrich.iterrows() ]\n",
    "DF_module_enrich"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SubtiWiki categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubtiWiki</th>\n",
       "      <th>BSU_number</th>\n",
       "      <th>FuncId</th>\n",
       "      <th>FuncName1</th>\n",
       "      <th>FuncName2</th>\n",
       "      <th>FuncName3</th>\n",
       "      <th>FuncName4</th>\n",
       "      <th>FuncName5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dat</td>\n",
       "      <td>BSU_09670</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ddl</td>\n",
       "      <td>BSU_04560</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>glmM</td>\n",
       "      <td>BSU_01770</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>glmS</td>\n",
       "      <td>BSU_01780</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mraY</td>\n",
       "      <td>BSU_15190</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SubtiWiki BSU_number  FuncId           FuncName1  \\\n",
       "0       dat  BSU_09670  SW 1.1  Cellular processes   \n",
       "1       ddl  BSU_04560  SW 1.1  Cellular processes   \n",
       "2      glmM  BSU_01770  SW 1.1  Cellular processes   \n",
       "3      glmS  BSU_01780  SW 1.1  Cellular processes   \n",
       "4      mraY  BSU_15190  SW 1.1  Cellular processes   \n",
       "\n",
       "                         FuncName2            FuncName3  \\\n",
       "0  cell envelope and cell division  cell wall synthesis   \n",
       "1  cell envelope and cell division  cell wall synthesis   \n",
       "2  cell envelope and cell division  cell wall synthesis   \n",
       "3  cell envelope and cell division  cell wall synthesis   \n",
       "4  cell envelope and cell division  cell wall synthesis   \n",
       "\n",
       "                       FuncName4 FuncName5  \n",
       "0  biosynthesis of peptidoglycan       NaN  \n",
       "1  biosynthesis of peptidoglycan       NaN  \n",
       "2  biosynthesis of peptidoglycan       NaN  \n",
       "3  biosynthesis of peptidoglycan       NaN  \n",
       "4  biosynthesis of peptidoglycan       NaN  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_subtiwiki = pd.read_csv(path.join(external_data,'subtiwiki_categories.csv'))\n",
    "DF_subtiwiki.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BSU_09670</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BSU_04560</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BSU_01770</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BSU_01780</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BSU_15190</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id   variable               value\n",
       "0  BSU_09670  FuncName1  Cellular processes\n",
       "1  BSU_04560  FuncName1  Cellular processes\n",
       "2  BSU_01770  FuncName1  Cellular processes\n",
       "3  BSU_01780  FuncName1  Cellular processes\n",
       "4  BSU_15190  FuncName1  Cellular processes"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Change the subtiwiki annotation format into a list of genes and categories\n",
    "DF_subtiwiki = DF_subtiwiki.rename({'BSU_number':'gene_id'},axis=1)\n",
    "DF_subtiwiki = DF_subtiwiki.melt(id_vars='gene_id',value_vars=['FuncName1','FuncName2','FuncName3','FuncName4','FuncName5'])\n",
    "DF_subtiwiki = DF_subtiwiki[DF_subtiwiki.value.notnull() & DF_subtiwiki.gene_id.isin(ica_data.gene_names)]\n",
    "DF_subtiwiki.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_subti_enrich = ica_data.compute_annotation_enrichment(DF_subtiwiki,'value')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>annotation</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>source</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>module_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>structural constituent of ribosome</td>\n",
       "      <td>4.033821e-29</td>\n",
       "      <td>3.513458e-26</td>\n",
       "      <td>0.313433</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.407767</td>\n",
       "      <td>21.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>cytosolic large ribosomal subunit</td>\n",
       "      <td>1.095099e-21</td>\n",
       "      <td>4.769155e-19</td>\n",
       "      <td>0.208955</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.325581</td>\n",
       "      <td>14.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>cytosolic small ribosomal subunit</td>\n",
       "      <td>3.988856e-07</td>\n",
       "      <td>1.158098e-04</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.129870</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>translation</td>\n",
       "      <td>1.218767e-06</td>\n",
       "      <td>2.653865e-04</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>0.126582</td>\n",
       "      <td>5.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>RNA binding</td>\n",
       "      <td>5.577349e-05</td>\n",
       "      <td>9.715741e-03</td>\n",
       "      <td>0.044776</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>cytoplasmic translation</td>\n",
       "      <td>3.248506e-04</td>\n",
       "      <td>4.042070e-02</td>\n",
       "      <td>0.029851</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.057971</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>large ribosomal subunit rRNA binding</td>\n",
       "      <td>3.248506e-04</td>\n",
       "      <td>4.042070e-02</td>\n",
       "      <td>0.029851</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.057971</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>17</td>\n",
       "      <td>plasma membrane-derived thylakoid photosystem II</td>\n",
       "      <td>6.278898e-05</td>\n",
       "      <td>5.468920e-02</td>\n",
       "      <td>0.059701</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>map04141</td>\n",
       "      <td>2.962003e-04</td>\n",
       "      <td>6.249826e-02</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>map03010</td>\n",
       "      <td>1.563306e-33</td>\n",
       "      <td>3.298575e-31</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.490566</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>26.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Ribosome</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>map02010</td>\n",
       "      <td>5.167328e-09</td>\n",
       "      <td>1.090306e-06</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.160714</td>\n",
       "      <td>9.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>ABC transporters</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>map02010</td>\n",
       "      <td>6.517585e-06</td>\n",
       "      <td>1.375210e-03</td>\n",
       "      <td>0.212121</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.119658</td>\n",
       "      <td>7.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>ABC transporters</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>map00196</td>\n",
       "      <td>2.623274e-05</td>\n",
       "      <td>5.535109e-03</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.090090</td>\n",
       "      <td>5.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Photosynthesis - antenna proteins</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>14</td>\n",
       "      <td>map00910</td>\n",
       "      <td>3.164213e-10</td>\n",
       "      <td>6.676489e-08</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.292683</td>\n",
       "      <td>6.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Nitrogen metabolism</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>14</td>\n",
       "      <td>map02010</td>\n",
       "      <td>2.448962e-04</td>\n",
       "      <td>1.498222e-02</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>0.059524</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>5.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>ABC transporters</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>14</td>\n",
       "      <td>map04724</td>\n",
       "      <td>1.426302e-04</td>\n",
       "      <td>1.498222e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.137931</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Glutamatergic synapse</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>14</td>\n",
       "      <td>map04727</td>\n",
       "      <td>2.840232e-04</td>\n",
       "      <td>1.498222e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>GABAergic synapse</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>14</td>\n",
       "      <td>map04217</td>\n",
       "      <td>4.713197e-04</td>\n",
       "      <td>1.988969e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Necroptosis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>15</td>\n",
       "      <td>map00195</td>\n",
       "      <td>1.372258e-04</td>\n",
       "      <td>2.895464e-02</td>\n",
       "      <td>0.053571</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>0.085106</td>\n",
       "      <td>12.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>224.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>17</td>\n",
       "      <td>map00195</td>\n",
       "      <td>1.855171e-23</td>\n",
       "      <td>3.914412e-21</td>\n",
       "      <td>0.313433</td>\n",
       "      <td>0.362069</td>\n",
       "      <td>0.336000</td>\n",
       "      <td>21.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>17</td>\n",
       "      <td>map00196</td>\n",
       "      <td>5.444976e-15</td>\n",
       "      <td>5.744450e-13</td>\n",
       "      <td>0.149254</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.243902</td>\n",
       "      <td>10.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Photosynthesis - antenna proteins</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>17</td>\n",
       "      <td>map01100</td>\n",
       "      <td>6.931914e-12</td>\n",
       "      <td>4.875446e-10</td>\n",
       "      <td>0.507463</td>\n",
       "      <td>0.060391</td>\n",
       "      <td>0.107937</td>\n",
       "      <td>34.0</td>\n",
       "      <td>563.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Metabolic pathways</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>17</td>\n",
       "      <td>map00710</td>\n",
       "      <td>2.509130e-04</td>\n",
       "      <td>1.323566e-02</td>\n",
       "      <td>0.059701</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.094118</td>\n",
       "      <td>4.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG pathways</td>\n",
       "      <td>Carbon fixation in photosynthetic organisms</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M00178</td>\n",
       "      <td>1.563306e-33</td>\n",
       "      <td>3.361107e-31</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.490566</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>26.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ribosome, bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>M00179</td>\n",
       "      <td>1.114389e-31</td>\n",
       "      <td>1.197968e-29</td>\n",
       "      <td>0.313433</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>0.432990</td>\n",
       "      <td>21.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ribosome, archaea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>M00222</td>\n",
       "      <td>1.087344e-18</td>\n",
       "      <td>2.337790e-16</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.818182</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Phosphate transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>M00316</td>\n",
       "      <td>4.350375e-08</td>\n",
       "      <td>9.353306e-06</td>\n",
       "      <td>0.214286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Manganese transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>M00246</td>\n",
       "      <td>5.352880e-05</td>\n",
       "      <td>1.150869e-02</td>\n",
       "      <td>0.012618</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.024922</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>317.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nickel transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>M00245</td>\n",
       "      <td>6.285372e-04</td>\n",
       "      <td>6.756775e-02</td>\n",
       "      <td>0.009464</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.018750</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>317.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cobalt/nickel transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9</td>\n",
       "      <td>M00240</td>\n",
       "      <td>2.636184e-08</td>\n",
       "      <td>5.667795e-06</td>\n",
       "      <td>0.121212</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.210526</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Iron complex transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9</td>\n",
       "      <td>M00190</td>\n",
       "      <td>6.441446e-06</td>\n",
       "      <td>6.924554e-04</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Iron(III) transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>13</td>\n",
       "      <td>M00155</td>\n",
       "      <td>3.225298e-04</td>\n",
       "      <td>6.934391e-02</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cytochrome c oxidase, prokaryotes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13</td>\n",
       "      <td>M00154</td>\n",
       "      <td>6.699769e-04</td>\n",
       "      <td>7.202251e-02</td>\n",
       "      <td>0.020833</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.040816</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cytochrome c oxidase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>13</td>\n",
       "      <td>M00004</td>\n",
       "      <td>1.279511e-03</td>\n",
       "      <td>9.169826e-02</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pentose phosphate pathway (Pentose phosphate c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>14</td>\n",
       "      <td>M00323</td>\n",
       "      <td>3.078404e-06</td>\n",
       "      <td>6.618568e-04</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.193548</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Urea transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>14</td>\n",
       "      <td>M00438</td>\n",
       "      <td>2.840232e-04</td>\n",
       "      <td>3.053249e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nitrate/nitrite transport system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>17</td>\n",
       "      <td>M00161</td>\n",
       "      <td>2.299900e-09</td>\n",
       "      <td>4.944786e-07</td>\n",
       "      <td>0.089552</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Photosystem II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>17</td>\n",
       "      <td>M00163</td>\n",
       "      <td>1.004887e-08</td>\n",
       "      <td>1.080253e-06</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.136986</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Photosystem I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>17</td>\n",
       "      <td>M00165</td>\n",
       "      <td>6.278898e-05</td>\n",
       "      <td>4.499877e-03</td>\n",
       "      <td>0.059701</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>19</td>\n",
       "      <td>M00316</td>\n",
       "      <td>1.741267e-04</td>\n",
       "      <td>3.743725e-02</td>\n",
       "      <td>0.014493</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>KEGG modules</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Manganese transport system</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    imodulon                                        annotation        pvalue  \\\n",
       "0          1                structural constituent of ribosome  4.033821e-29   \n",
       "1          1                 cytosolic large ribosomal subunit  1.095099e-21   \n",
       "2          1                 cytosolic small ribosomal subunit  3.988856e-07   \n",
       "3          1                                       translation  1.218767e-06   \n",
       "4          1                                       RNA binding  5.577349e-05   \n",
       "5          1                           cytoplasmic translation  3.248506e-04   \n",
       "6          1              large ribosomal subunit rRNA binding  3.248506e-04   \n",
       "7         17  plasma membrane-derived thylakoid photosystem II  6.278898e-05   \n",
       "0          0                                          map04141  2.962003e-04   \n",
       "1          1                                          map03010  1.563306e-33   \n",
       "2          3                                          map02010  5.167328e-09   \n",
       "3          9                                          map02010  6.517585e-06   \n",
       "4         13                                          map00196  2.623274e-05   \n",
       "5         14                                          map00910  3.164213e-10   \n",
       "6         14                                          map02010  2.448962e-04   \n",
       "7         14                                          map04724  1.426302e-04   \n",
       "8         14                                          map04727  2.840232e-04   \n",
       "9         14                                          map04217  4.713197e-04   \n",
       "10        15                                          map00195  1.372258e-04   \n",
       "11        17                                          map00195  1.855171e-23   \n",
       "12        17                                          map00196  5.444976e-15   \n",
       "13        17                                          map01100  6.931914e-12   \n",
       "14        17                                          map00710  2.509130e-04   \n",
       "0          1                                            M00178  1.563306e-33   \n",
       "1          1                                            M00179  1.114389e-31   \n",
       "2          3                                            M00222  1.087344e-18   \n",
       "3          4                                            M00316  4.350375e-08   \n",
       "4          5                                            M00246  5.352880e-05   \n",
       "5          5                                            M00245  6.285372e-04   \n",
       "6          9                                            M00240  2.636184e-08   \n",
       "7          9                                            M00190  6.441446e-06   \n",
       "8         13                                            M00155  3.225298e-04   \n",
       "9         13                                            M00154  6.699769e-04   \n",
       "10        13                                            M00004  1.279511e-03   \n",
       "11        14                                            M00323  3.078404e-06   \n",
       "12        14                                            M00438  2.840232e-04   \n",
       "13        17                                            M00161  2.299900e-09   \n",
       "14        17                                            M00163  1.004887e-08   \n",
       "15        17                                            M00165  6.278898e-05   \n",
       "16        19                                            M00316  1.741267e-04   \n",
       "\n",
       "          qvalue  precision    recall   f1score    TP  target_set_size  \\\n",
       "0   3.513458e-26   0.313433  0.583333  0.407767  21.0             36.0   \n",
       "1   4.769155e-19   0.208955  0.736842  0.325581  14.0             19.0   \n",
       "2   1.158098e-04   0.074627  0.500000  0.129870   5.0             10.0   \n",
       "3   2.653865e-04   0.074627  0.416667  0.126582   5.0             12.0   \n",
       "4   9.715741e-03   0.044776  0.600000  0.083333   3.0              5.0   \n",
       "5   4.042070e-02   0.029851  1.000000  0.057971   2.0              2.0   \n",
       "6   4.042070e-02   0.029851  1.000000  0.057971   2.0              2.0   \n",
       "7   5.468920e-02   0.059701  0.307692  0.100000   4.0             16.0   \n",
       "0   6.249826e-02   0.031250  1.000000  0.060606   2.0              2.0   \n",
       "1   3.298575e-31   0.388060  0.490566  0.433333  26.0             53.0   \n",
       "2   1.090306e-06   0.321429  0.107143  0.160714   9.0             84.0   \n",
       "3   1.375210e-03   0.212121  0.083333  0.119658   7.0             84.0   \n",
       "4   5.535109e-03   0.052083  0.333333  0.090090   5.0             15.0   \n",
       "5   6.676489e-08   0.230769  0.400000  0.292683   6.0             15.0   \n",
       "6   1.498222e-02   0.192308  0.059524  0.090909   5.0             84.0   \n",
       "7   1.498222e-02   0.076923  0.666667  0.137931   2.0              3.0   \n",
       "8   1.498222e-02   0.076923  0.500000  0.133333   2.0              4.0   \n",
       "9   1.988969e-02   0.076923  0.400000  0.129032   2.0              5.0   \n",
       "10  2.895464e-02   0.053571  0.206897  0.085106  12.0             58.0   \n",
       "11  3.914412e-21   0.313433  0.362069  0.336000  21.0             58.0   \n",
       "12  5.744450e-13   0.149254  0.666667  0.243902  10.0             15.0   \n",
       "13  4.875446e-10   0.507463  0.060391  0.107937  34.0            563.0   \n",
       "14  1.323566e-02   0.059701  0.222222  0.094118   4.0             18.0   \n",
       "0   3.361107e-31   0.388060  0.490566  0.433333  26.0             53.0   \n",
       "1   1.197968e-29   0.313433  0.700000  0.432990  21.0             30.0   \n",
       "2   2.337790e-16   0.321429  0.818182  0.461538   9.0             11.0   \n",
       "3   9.353306e-06   0.214286  1.000000  0.352941   3.0              3.0   \n",
       "4   1.150869e-02   0.012618  1.000000  0.024922   4.0              4.0   \n",
       "5   6.756775e-02   0.009464  1.000000  0.018750   3.0              3.0   \n",
       "6   5.667795e-06   0.121212  0.800000  0.210526   4.0              5.0   \n",
       "7   6.924554e-04   0.090909  0.600000  0.157895   3.0              5.0   \n",
       "8   6.934391e-02   0.031250  0.500000  0.058824   3.0              6.0   \n",
       "9   7.202251e-02   0.020833  1.000000  0.040816   2.0              2.0   \n",
       "10  9.169826e-02   0.031250  0.333333  0.057143   3.0              9.0   \n",
       "11  6.618568e-04   0.115385  0.600000  0.193548   3.0              5.0   \n",
       "12  3.053249e-02   0.076923  0.500000  0.133333   2.0              4.0   \n",
       "13  4.944786e-07   0.089552  0.666667  0.157895   6.0              9.0   \n",
       "14  1.080253e-06   0.074627  0.833333  0.136986   5.0              6.0   \n",
       "15  4.499877e-03   0.059701  0.307692  0.100000   4.0             13.0   \n",
       "16  3.743725e-02   0.014493  1.000000  0.028571   3.0              3.0   \n",
       "\n",
       "    imodulon_size         source                                 pathway_name  \\\n",
       "0            67.0             GO                                          NaN   \n",
       "1            67.0             GO                                          NaN   \n",
       "2            67.0             GO                                          NaN   \n",
       "3            67.0             GO                                          NaN   \n",
       "4            67.0             GO                                          NaN   \n",
       "5            67.0             GO                                          NaN   \n",
       "6            67.0             GO                                          NaN   \n",
       "7            67.0             GO                                          NaN   \n",
       "0            64.0  KEGG pathways  Protein processing in endoplasmic reticulum   \n",
       "1            67.0  KEGG pathways                                     Ribosome   \n",
       "2            28.0  KEGG pathways                             ABC transporters   \n",
       "3            33.0  KEGG pathways                             ABC transporters   \n",
       "4            96.0  KEGG pathways            Photosynthesis - antenna proteins   \n",
       "5            26.0  KEGG pathways                          Nitrogen metabolism   \n",
       "6            26.0  KEGG pathways                             ABC transporters   \n",
       "7            26.0  KEGG pathways                        Glutamatergic synapse   \n",
       "8            26.0  KEGG pathways                            GABAergic synapse   \n",
       "9            26.0  KEGG pathways                                  Necroptosis   \n",
       "10          224.0  KEGG pathways                               Photosynthesis   \n",
       "11           67.0  KEGG pathways                               Photosynthesis   \n",
       "12           67.0  KEGG pathways            Photosynthesis - antenna proteins   \n",
       "13           67.0  KEGG pathways                           Metabolic pathways   \n",
       "14           67.0  KEGG pathways  Carbon fixation in photosynthetic organisms   \n",
       "0            67.0   KEGG modules                                          NaN   \n",
       "1            67.0   KEGG modules                                          NaN   \n",
       "2            28.0   KEGG modules                                          NaN   \n",
       "3            14.0   KEGG modules                                          NaN   \n",
       "4           317.0   KEGG modules                                          NaN   \n",
       "5           317.0   KEGG modules                                          NaN   \n",
       "6            33.0   KEGG modules                                          NaN   \n",
       "7            33.0   KEGG modules                                          NaN   \n",
       "8            96.0   KEGG modules                                          NaN   \n",
       "9            96.0   KEGG modules                                          NaN   \n",
       "10           96.0   KEGG modules                                          NaN   \n",
       "11           26.0   KEGG modules                                          NaN   \n",
       "12           26.0   KEGG modules                                          NaN   \n",
       "13           67.0   KEGG modules                                          NaN   \n",
       "14           67.0   KEGG modules                                          NaN   \n",
       "15           67.0   KEGG modules                                          NaN   \n",
       "16          207.0   KEGG modules                                          NaN   \n",
       "\n",
       "                                          module_name  \n",
       "0                                                 NaN  \n",
       "1                                                 NaN  \n",
       "2                                                 NaN  \n",
       "3                                                 NaN  \n",
       "4                                                 NaN  \n",
       "5                                                 NaN  \n",
       "6                                                 NaN  \n",
       "7                                                 NaN  \n",
       "0                                                 NaN  \n",
       "1                                                 NaN  \n",
       "2                                                 NaN  \n",
       "3                                                 NaN  \n",
       "4                                                 NaN  \n",
       "5                                                 NaN  \n",
       "6                                                 NaN  \n",
       "7                                                 NaN  \n",
       "8                                                 NaN  \n",
       "9                                                 NaN  \n",
       "10                                                NaN  \n",
       "11                                                NaN  \n",
       "12                                                NaN  \n",
       "13                                                NaN  \n",
       "14                                                NaN  \n",
       "0                                  Ribosome, bacteria  \n",
       "1                                   Ribosome, archaea  \n",
       "2                          Phosphate transport system  \n",
       "3                          Manganese transport system  \n",
       "4                             Nickel transport system  \n",
       "5                      Cobalt/nickel transport system  \n",
       "6                       Iron complex transport system  \n",
       "7                          Iron(III) transport system  \n",
       "8                   Cytochrome c oxidase, prokaryotes  \n",
       "9                                Cytochrome c oxidase  \n",
       "10  Pentose phosphate pathway (Pentose phosphate c...  \n",
       "11                              Urea transport system  \n",
       "12                   Nitrate/nitrite transport system  \n",
       "13                                     Photosystem II  \n",
       "14                                      Photosystem I  \n",
       "15   Reductive pentose phosphate cycle (Calvin cycle)  \n",
       "16                         Manganese transport system  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DF_GO_enrich['source'] = 'GO'\n",
    "DF_pathway_enrich['source'] = 'KEGG pathways'\n",
    "DF_module_enrich['source'] = 'KEGG modules'\n",
    "# DF_subti_enrich['source'] = 'SubtiWiki'\n",
    "\n",
    "DF_GO_enrich.rename({'gene_ontology':'annotation'},axis=1, inplace=True)\n",
    "DF_pathway_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "DF_module_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "# DF_subti_enrich.rename({'value':'annotation'},axis=1, inplace=True)\n",
    "\n",
    "DF_enrichments = pd.concat([DF_GO_enrich, DF_pathway_enrich, DF_module_enrich])\n",
    "display(DF_enrichments)\n",
    "DF_enrichments.to_csv(path.join(data_dir,'functional_enrichments.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for single gene iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some iModulons are dominated by a single, high-coefficient gene. These iModulons may result from:\n",
    "1. Overdecomposition of the dataset to identify noisy genes\n",
    "1. Artificial knock-out of single genes\n",
    "1. Regulons with only one target gene\n",
    "\n",
    "No matter what causes these iModulons, it is important to be aware of them. The find_single_gene_imodulons function identifies iModulons that are likely dominated by a single gene.\n",
    "\n",
    "The iModulons identified by ``find_single_gene_imodulons`` may contain more than one gene, since a threshold-agnostic method is used to identify these iModulons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg_imods = ica_data.find_single_gene_imodulons(save=True)\n",
    "len(sg_imods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,mod in enumerate(sg_imods):\n",
    "    ica_data.rename_imodulons({mod:'SG_'+str(i+1)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'single_gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_77644/1381282101.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msingle_gene\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'single_gene'"
     ]
    }
   ],
   "source": [
    "ica_data.imodulon_table[ica_data.imodulon_table.single_gene == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'SG_1'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3360\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3361\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.index.Int64Engine._check_type\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.index.Int64Engine._check_type\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'SG_1'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_77644/1542687927.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mview_imodulon\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'SG_1'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pymodulon/core.py\u001b[0m in \u001b[0;36mview_imodulon\u001b[0;34m(self, imodulon)\u001b[0m\n\u001b[1;32m    551\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    552\u001b[0m         \u001b[0;31m# Find genes in iModulon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 553\u001b[0;31m         \u001b[0min_imodulon\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mabs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mM\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mimodulon\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mthresholds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mimodulon\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    554\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    555\u001b[0m         \u001b[0;31m# Get gene weights information\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3456\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3457\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3458\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3459\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3460\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/modulome/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3361\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3363\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3364\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3365\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhasnans\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'SG_1'"
     ]
    }
   ],
   "source": [
    "ica_data.view_imodulon('SG_1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save iModulon object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.util import explained_variance\n",
    "from pymodulon.io import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add iModulon sizes and explained variance\n",
    "for im in ica_data.imodulon_names:\n",
    "    ica_data.imodulon_table.loc[im,'imodulon_size'] = len(ica_data.view_imodulon(im))\n",
    "    ica_data.imodulon_table.loc[im,'explained_variance'] = explained_variance(ica_data,imodulons=im)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>explained_variance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>64.0</td>\n",
       "      <td>0.051002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>67.0</td>\n",
       "      <td>0.022770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>174.0</td>\n",
       "      <td>0.003296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>28.0</td>\n",
       "      <td>0.006637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14.0</td>\n",
       "      <td>0.005571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>317.0</td>\n",
       "      <td>0.022298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.033855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>72.0</td>\n",
       "      <td>0.014793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>18.0</td>\n",
       "      <td>0.024013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>33.0</td>\n",
       "      <td>0.026245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.039865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>136.0</td>\n",
       "      <td>0.038922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>68.0</td>\n",
       "      <td>0.010907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>96.0</td>\n",
       "      <td>0.032078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>26.0</td>\n",
       "      <td>0.009535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>224.0</td>\n",
       "      <td>0.004825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>123.0</td>\n",
       "      <td>0.024131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>67.0</td>\n",
       "      <td>0.027164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>10.0</td>\n",
       "      <td>0.020790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>207.0</td>\n",
       "      <td>0.085741</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    imodulon_size  explained_variance\n",
       "0            64.0            0.051002\n",
       "1            67.0            0.022770\n",
       "2           174.0            0.003296\n",
       "3            28.0            0.006637\n",
       "4            14.0            0.005571\n",
       "5           317.0            0.022298\n",
       "6             4.0            0.033855\n",
       "7            72.0            0.014793\n",
       "8            18.0            0.024013\n",
       "9            33.0            0.026245\n",
       "10           15.0            0.039865\n",
       "11          136.0            0.038922\n",
       "12           68.0            0.010907\n",
       "13           96.0            0.032078\n",
       "14           26.0            0.009535\n",
       "15          224.0            0.004825\n",
       "16          123.0            0.024131\n",
       "17           67.0            0.027164\n",
       "18           10.0            0.020790\n",
       "19          207.0            0.085741"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will save your iModulon table, your thresholds, and any other information stored in the ica_data object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data, path.join(data_dir,'bsu_raw.json.gz'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you prefer to view and edit your iModulon table in excel, save it as a CSV and reload the iModulon as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.imodulon_table.to_csv(path.join(data_dir,'imodulon_table_raw.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "modulome",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "vscode": {
   "interpreter": {
    "hash": "ef87fd4debf949ec83da72b2a651a3b92f1ee0ca5a38d6d1c1e00db766f058a0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
